src/Orleans/Threading/CancellationSourcesExtension.cs

```diff
@@ -0,0 +1,52 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Orleans.Concurrency;
+using Orleans.Runtime;
+
+namespace Orleans.Threading
+{
+    internal class CancellationSourcesExtension : ICancellationSourcesExtension
+    {
+        private readonly Lazy<TraceLogger> _logger = new Lazy<TraceLogger>(() =>
+            TraceLogger.GetLogger("CancellationSourcesExtension", TraceLogger.LoggerType.Application));
+        private readonly Dictionary<Guid, CancellationTokenSource> _cancellationTokenSources;
+
+        public CancellationSourcesExtension()
+        {
+            _cancellationTokenSources = new Dictionary<Guid, CancellationTokenSource>();
+        }
+
+        public Task CancelTokenSource(Guid tokenId)
+        {
+            CancellationTokenSource cts;
+            if (!_cancellationTokenSources.TryGetValue(tokenId, out cts))
+            {
+                if (_logger.Value.IsWarning)
+                {
+                    _logger.Value.Warn(ErrorCode.CancellationTokenCancelFailed, "Remote token cancellation failed: token was not found");
+                }
+
+                return TaskDone.Done;
+            }
+
+            cts.Cancel();
+            return TaskDone.Done;
+        }
+
+        internal CancellationTokenSource CreateCancellationTokenSource(Guid tokenId)
```

Comment:
The token source is guaranteed to be returned, so I've changed to `GetOrCreateCancellationTokenSource`